server/src/main/java/org/elasticsearch/cluster/coordination/JoinHelper.java

```diff
@@ -0,0 +1,247 @@
+
+            @Override
+            public ClusterTasksResult<JoinTaskExecutor.Task> execute(ClusterState currentState, List<JoinTaskExecutor.Task> joiningTasks)
+                throws Exception {
+                // This is called when preparing the next cluster state for publication. There is no guarantee that the term we see here is
+                // the term under which this state will eventually be published: the current term may be increased after this check due to
+                // some other activity. That the term is correct is, however, checked properly during publication, so it is sufficient to
+                // check it here on a best-effort basis. This is fine because a concurrent change indicates the existence of another leader
+                // in a higher term which will cause this node to stand down.
+
+                final long currentTerm = currentTermSupplier.getAsLong();
+                if (currentState.term() != currentTerm) {
+                    currentState = ClusterState.builder(currentState).term(currentTerm).build();
+                }
+                return super.execute(currentState, joiningTasks);
+            }
+
+        };
+
+        transportService.registerRequestHandler(JOIN_ACTION_NAME, ThreadPool.Names.GENERIC, false, false, JoinRequest::new,
+            (request, channel, task) -> handleJoinRequest(request, new JoinCallback() {
+
+                @Override
+                public void onSuccess() {
+                    try {
+                        channel.sendResponse(TransportResponse.Empty.INSTANCE);
+                    } catch (IOException e) {
+                        onFailure(e);
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    try {
+                        channel.sendResponse(e);
+                    } catch (Exception inner) {
+                        inner.addSuppressed(e);
+                        logger.warn("failed to send back failure on join request", inner);
+                    }
+                }
+
+                @Override
+                public String toString() {
+                    return "JoinCallback{request=" + request + "}";
+                }
+            }));
+    }
+
+    private void handleJoinRequest(JoinRequest joinRequest, JoinCallback joinCallback) {
+        transportService.connectToNode(joinRequest.getSourceNode());
+
+        final Optional<Join> optionalJoin = joinRequest.getOptionalJoin();
+        final boolean justBecameLeader = optionalJoin.isPresent() && joinHandler.test(optionalJoin.get());
+
+        synchronized (mutex) {
+            joinAccumulator.handleJoinRequest(joinRequest.getSourceNode(), joinCallback);
+
+            if (justBecameLeader) {
+                joinAccumulator.submitPendingJoins();
+                joinAccumulator = new LeaderJoinAccumulator();
```

Comment:
I see a risk here that we become leader, then become candidate, and _then_ overwrite the accumulator here. It's ok if we `becomeCandidate` on each new election, but this is at least worthy of a comment.

Reply:
The `Coordinator` becomes leader in `joinHandler.test()` not in `handleJoinRequest`, and that's outside this mutex, so it's technically possible that it could become a candidate again before this synchronised block.

Reply:
oh, right. Should we extend the mutex?

Reply:
I think this race here can also lead to another odd situation:
Assume you have two concurrent join requests, both are required to become leader. The first that enters handleJoin or Coordinator will return becomeLeader = false, the second one will return true. If the second one now gets to execute this section here first, it will be submitted to the masterservice whereas the second will only be submitted in a follow-up. This means its possible that only the second node will be part of the cluster state that is published as becoming leader.